### PR TITLE
Ownable-PSP1155 example

### PR DESCRIPTION
**Pull request by PierreOssun**

### CODE REVIEW

File: contracts/token/psp1155/traits.rs [hunk 1]
```diff
@@ -64,6 +64,17 @@ pub trait IPSP1155 {
     );
 }
 
+#[brush::trait_definition]
+pub trait IPSP1155Mint {
+    /// Creates a new token.
```
Comment (VargSupercolony) [inline contracts/token/psp1155/traits.rs:69]: I guess it doesn't really create a new token per say, it just 'mints' it (increases the emission), just like 'burn' below doesn't delete the token, just decreases the emission/supply

File: tests/ownable.tests.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    // Arrange - Sender mint 100 of token 0 and 100 of token 1
+    await expect(contract.tx.mint(sender.address, bnArg(0), 100)).to.eventually.be.fulfilled
+    await expect(contract.tx.mint(sender.address, bnArg(1), 100)).to.eventually.be.fulfilled
+
+    // Act - Sender transfer 20 of token 0 and 70 of token 1
+    await expect(contract.tx.safeBatchTransferFrom(sender.address, alice.address, [bnArg(0), bnArg(1)], [20, 70], 'data')).to.eventually.be.fulfilled
+
+    // Assert - Alice own 20 of token 0 and own 70 0f token 1
+    await expect(query.balanceOf(alice.address, bnArg(0))).to.have.output(20)
```
Comment (xgreenx) [inline tests/ownable.tests.ts:78]: You can check it via batch balance of

File: tests/ownable.tests.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
+import { consts } from './constants'
+import {bnArg, expect, fromSigner, setupContract} from './helpers'
+
+describe('MY_OWNABLE', () => {
+  async function setup() {
+    return setupContract('my_ownable', 'new')
+  }
+
+  it('PSP 1155 - mint works', async () => {
```
Comment (xgreenx) [inline tests/ownable.tests.ts:9]: You also need to verify `is_approved_for_all`, `set_approval_for_all`, and both "safe" functions(for that you need to deploy a mock Receiver and verify that on_received was called).

About safe methods. Could you also add the same test for PSP721 please?

**Review by xgreenx — CHANGES_REQUESTED** [review]

**Review by VargSupercolony — COMMENTED** [review]
clear for me, change according to Green and my approval is there

File: examples/ownable/lib.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    impl Ownable for PSP1155Struct {}
+    impl PSP1155 for PSP1155Struct {}
+
+    impl PSP1155Mint for PSP1155Struct {}
+    impl IPSP1155Mint for PSP1155Struct {
+        #[ink(message)]
+        fn mint(&mut self, to: AccountId, id: Id, amount: Balance) {
+            self.only_owner();
```
Comment (xgreenx) [inline examples/ownable/lib.rs:52]: You can use modifiers macro here

File: tests/access-control.tests.ts [hunk 5]
```diff
@@ -57,12 +61,39 @@ describe('MY_ACCESS_CONTROL', () => {
...
     // Act - Alice transfer the token form sender to bob
     await expect(fromSigner(contract, alice.address).tx.transferFrom(sender.address, bob.address, bnArg(0)))
-        .to.eventually.be.fulfilled
+      .to.eventually.be.fulfilled
 
     // Assert - Bob is now owner of the token
     await expect(query.ownerOf(bnArg(0))).to.have.output(bob.address)
   })
 
+  it('PSP 721 - safe transfer works', async () => {
```
Comment (xgreenx) [inline tests/access-control.tests.ts:70]: Will be cool to test the scenario where the Receiver will reject the transfer to verify that it works properly.

For that, you can add an additional method to mock contract `revert_next_transfer()`

**Review by xgreenx — APPROVED** [review]

File: mock/psp1155-receiver/lib.rs [hunk 6]
```diff
@@ -9,33 +9,47 @@ pub mod erc1155_receiver {
...
             self.call_counter += 1;
+            self.revert_next_transfer = false;
             Ok(())
         }
 
         #[ink(message)]
         fn on_psp1155_batch_received(&mut self, _operator: AccountId, _from: AccountId,
                                          _ids: Vec<Id>, _values: Vec<Balance>, _data: Vec<u8>) -> Result<(), PSP1155ReceiverError> {
+            if self.revert_next_transfer {
+                return Err(PSP1155ReceiverError::TransferRejected)
```
Comment (xgreenx) [inline mock/psp1155-receiver/lib.rs:49]: You need to do it `self.revert_next_transfer = false;` here=)

File: contracts/token/psp721/traits.rs [hunk 7]
```diff
@@ -81,7 +81,7 @@ pub trait IPSP721Mint {
 #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
 pub enum PSP721ReceiverError {
     /// Returned if transfer is rejected.
-    TransferRejected(String),
+    TransferRejected,
```
Comment (xgreenx) [inline contracts/token/psp721/traits.rs:84]: Please return back the string=)

**Review by xgreenx — CHANGES_REQUESTED** [review]

File: mock/psp1155-receiver/lib.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        #[ink(message)]
+        fn on_psp1155_batch_received(&mut self, _operator: AccountId, _from: AccountId,
+                                         _ids: Vec<Id>, _values: Vec<Balance>, _data: Vec<u8>) -> Result<(), PSP1155ReceiverError> {
+            if self.revert_next_transfer {
+                self.revert_next_transfer = false;
+                return Err(PSP1155ReceiverError::TransferRejected)
+            }
+            self.call_counter += 1;
+            self.revert_next_transfer = false;
```
Comment (xgreenx) [inline mock/psp1155-receiver/lib.rs:54]: You can remove this line and in PSP721 too=)

**Review by xgreenx — APPROVED** [review]

File: contracts/token/psp1155/traits.rs [hunk 9]
```diff
@@ -66,7 +66,18 @@ pub trait IPSP1155 {
...
+    fn mint(&mut self, to: AccountId, id: Id, amount: Balance);
+
+    /// Burns an existing token. Only the owner can burn the token.
+    #[ink(message)]
+    fn burn(&mut self, from: AccountId, id: Id, amount: Balance);
+}
+
+#[brush::trait_definition]
+pub trait IPSP1155MetadataURI {
     /// Returns the URI for token type `id`.
```
[the remark this answers is not in the document]
Comment (PierreOssun) [reply contracts/token/psp1155/traits.rs:81]: I was wondering if we should remove both or add it in erc1155. Adding it !

**Review by xgreenx — APPROVED** [review]

**Pull request merged**